sw/device/tests/i2c_target_test.c

```diff
@@ -49,14 +55,94 @@ status_t configure_device_address(ujson_t *uj, dif_i2c_t *i2c) {
   return RESP_OK_STATUS(uj);
 }
 
-status_t command_processor(ujson_t *uj) {
+static status_t wait_for_acq_fifo(dif_i2c_t *i2c, uint8_t at_least,
+                                  const ibex_timeout_t *deadline) {
+  uint8_t acq_fifo_lvl;
+  while (!ibex_timeout_check(deadline)) {
+    if (dif_i2c_get_fifo_levels(i2c, NULL, NULL, NULL, &acq_fifo_lvl) ==
+            kDifOk &&
+        acq_fifo_lvl >= at_least) {
+      return OK_STATUS();
+    }
+  }
+  return DEADLINE_EXCEEDED();
+}
+
+static status_t recv_write_transaction(dif_i2c_t *i2c, i2c_transaction_t *txn,
+                                       uint32_t micros, uint32_t delay_micros) {
+  ibex_timeout_t deadline = ibex_timeout_init(micros);
+  uint8_t byte;
+  dif_i2c_signal_t signal;
+
+  // Address phase.
+  TRY(wait_for_acq_fifo(i2c, 1, &deadline));
+  TRY(dif_i2c_acquire_byte(i2c, &byte, &signal));
+  TRY_CHECK(signal == kDifI2cSignalStart);
+  txn->address = byte >> 1;
+
+  // Data phase.
+  while (true) {
+    if (txn->length % 64 == 0 && delay_micros) {
+      busy_spin_micros(delay_micros);
+    }
+    TRY(wait_for_acq_fifo(i2c, 1, &deadline));
+    TRY(dif_i2c_acquire_byte(i2c, &byte, &signal));
+    if (signal != kDifI2cSignalNone) {
+      break;
+    }
+    txn->data[txn->length++] = byte;
```

Comment:
Should we check the array boundaries here?

Reply:
No need: `length` is a uint8_t and can never overflow `data`.